homeassistant/components/mqtt/discovery.py

```diff
@@ -65,6 +67,16 @@ def async_device_message_received(topic, payload, qos):
             payload[CONF_STATE_TOPIC] = '{}/{}/{}/state'.format(
                 discovery_topic, component, object_id)
 
+        discovery_hash = json.dumps([component, object_id], sort_keys=True)
```

Comment:
No need to JSON dump it, strings are immutable and so if you make it a tuple `(component, object_id)`, it's hashable.